src/gl/value.js

```diff
@@ -28,920 +27,434 @@ export interface Value<T> {
+        gl.pixelStorei(gl.UNPACK_ALIGNMENT, v);
     }
 }
 
-export class PixelStoreUnpackPremultiplyAlpha implements Value<boolean> {
-    context: Context;
-    current: boolean;
-    default: boolean;
-    dirty: boolean;
-
-    constructor(context: Context) {
-        this.context = context;
-        this.default = false;
-        this.current = this.default;
-        this.dirty = false;
+export class PixelStoreUnpackPremultiplyAlpha extends BaseValue<boolean> {
+    getDefault(): boolean {
+        return false;
     }
-
-    get(): boolean { return this.current; }
-
-    setDefault(): void { this.set(this.default); }
-
-    set(v: boolean): void {
-        if (this.current !== v || this.dirty === true) {
-            const gl = this.context.gl;
-            gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, (v: any));
-            this.current = v;
-            this.dirty = false;
-        }
+    _set(v: boolean): void {
+        const gl = this.gl;
+        gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, (v: any));
     }
 }
 
-export class PixelStoreUnpackFlipY implements Value<boolean> {
-    context: Context;
-    current: boolean;
-    default: boolean;
-    dirty: boolean;
-
-    constructor(context: Context) {
-        this.context = context;
-        this.default = false;
-        this.current = this.default;
-        this.dirty = false;
+export class PixelStoreUnpackFlipY extends BaseValue<boolean> {
+    getDefault(): boolean {
+        return false;
     }
-
-    get(): boolean { return this.current; }
-
-    setDefault(): void { this.set(this.default); }
-
-    set(v: boolean): void {
-        if (this.current !== v || this.dirty === true) {
-            const gl = this.context.gl;
-            gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, (v: any));
-            this.current = v;
-            this.dirty = false;
-        }
+    _set(v: boolean): void {
+        const gl = this.gl;
+        gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, (v: any));
     }
 }
 
-/**
- * Framebuffer values
- * @private
- */
-export class FramebufferValue<T> {
-    context: Context;
+class FramebufferAttachment<T> extends BaseValue<?T> {
     parent: WebGLFramebuffer;
-    current: ?T;
-    default: ?T;
-    dirty: boolean;
+    context: Context;
 
     constructor(context: Context, parent: WebGLFramebuffer) {
+        super(context);
         this.context = context;
-        this.default = null;
-        this.current = this.default;
-        this.dirty = false;
         this.parent = parent;
     }
-
-    get(): ?T { return this.current; }
-}
-
-export class ColorAttachment extends FramebufferValue<?WebGLTexture> implements Value<?WebGLTexture> {
-    dirty: boolean;
-
-    constructor(context: Context, parent: WebGLFramebuffer) {
-        super(context, parent);
-        this.dirty = false;
-    }
-
-    setDefault(): void { this.set(this.default); }
-
-    set(v: ?WebGLTexture): void {
-        if (this.dirty || this.current !== v) {
```

Comment:
Was there a reason that the dirty/inequality check order was reversed here before? I'm not actually sure what the equality operator for `WebGLTexture` does although it seems unlikely to make any difference that we can now call it more frequently (when the `ColorAttachment` is dirty).

Reply:
I think the base class `set` function does the dirty/inequality check before calling each child class' `_set` function.

Reply:
I think there's no difference — comparing boolean shouldn't be slower than comparing object references.

Reply:
ooh I missed the key word `reversed` – nvm!